### PR TITLE
Fetch

### PR DESCRIPTION
**Pull request by tmcdonnell87**
Not sure why it's also trying to merge 64546ab in as that was already merged. This was pushed a few days ago but got pulled into something before I created a PR.

### CODE REVIEW

File: db_processing/core/utils/file_utils.py [hunk 1]
```diff
@@ -1,76 +1,110 @@
...
+    stream.seek(os.SEEK_SET)
+    return stream, filename
+
+def decrypt_file(stream, key_data, passphrase):
+    gpg = gnupg.GPG()
+    try:
+        gpg.import_keys(key_data)
+        decrypted = gpg.decrypt_file(stream, passphrase=passphrase)
+    except Exception as exc:
+        stream.close()
```
Comment (GrivIN) [inline db_processing/core/utils/file_utils.py]: I do not like this design, We open stream in one function and close it in another... It took me too long to figure out and check that file will be always closed. what about something like
```python
def read_file(data_filename ...):
    ...
    with get_s3_file(parsed) as stream:
        data = decrypt_file(stream, key_data, passphrase)
```

File: db_processing/core/utils/file_utils.py [hunk 2]
```diff
@@ -1,76 +1,110 @@
...
+def get_s3_file(bucketname, path):
+    s3 = boto3.resource('s3')
+    try:
+        bucket = s3.Bucket(bucketname)
+        # s3 doesn't support glob
+        if path.startswith('/'):
+            path = path[1:]
+        prefix = re.split(r'\[|\*|\?', path)[0]
+        files = bucket.objects.filter(Prefix=prefix)
+        files = [f for f in files if fnmatch.fnmatch(f.key, path)]
```
Comment (GrivIN) [inline db_processing/core/utils/file_utils.py]: lets use generator here

```python
files = (f for f in files if fnmatch.fnmatch(f.key, path))
```

File: db_processing/core/utils/file_utils.py [hunk 3]
```diff
@@ -1,76 +1,110 @@
...
+        # if supplied with a prefix (e.g., folder such as 'MDP'), get the most recent file
+        key = [obj for obj in sorted(
```
Comment (GrivIN) [inline db_processing/core/utils/file_utils.py]: simply:
```python
key = sorted(
           files,
           key=lambda obj: int(obj.last_modified.strftime('%s')),
           reverse=True
)[0].key
```

File: db_processing/core/utils/file_utils.py [hunk 4]
```diff
@@ -1,76 +1,110 @@
...
+    log.info('status: %s', decrypted.status)
+    return decrypted.data.decode('utf-8')
+
+def get_file(path):
+    parsed = urlparse(path)
+    if parsed.scheme == 's3':
+        return get_s3_file(parsed.hostname, parsed.path)
+    elif parsed.scheme == 'sftp':
+        return get_sftp_file(parsed.hostname, parsed.path, parsed.username, parsed.password, parsed.port)
+    return get_local_file(parsed.path)
```
Comment (GrivIN) [inline db_processing/core/utils/file_utils.py]: lets use one convention:
```python
get_sftp_file(parsed)
get_s3_file(parsed)
et_local_file(parsed)
```

File: db_processing/core/utils/file_utils.py [hunk 5]
```diff
@@ -1,76 +1,110 @@
...
+        with pysftp.Connection(
+            hostname,
+            username=username,
+            password=password,
+            port=port,
+            cnopts=cnopts) as sftp:
+            sftp.chdir(dirname)
+            files = sftp.listdir_attr()
+            files = [f for f in files if fnmatch.fnmatch(f.filename, filepath)]
+            filename = [f for f in sorted(files, key=lambda x: x.st_mtime, reverse=True)][0].filename
```
Comment (GrivIN) [inline db_processing/core/utils/file_utils.py]:
```python
filename = sorted(files, key=lambda x: x.st_mtime, reverse=True)[0].filename
```

File: db_processing/core/utils/file_utils.py [hunk 6]
```diff
@@ -1,76 +1,110 @@
...
+    if SFTP_DISABLE_KNOWN_HOSTS:
+        cnopts.hostkeys = None
+    stream = BytesIO()
+    try:
+        with pysftp.Connection(
+            hostname,
+            username=username,
+            password=password,
+            port=port,
+            cnopts=cnopts) as sftp:
```
[the remark this answers is not in the document]
Comment (tmcdonnell87) [reply db_processing/core/utils/file_utils.py]: pylint disagrees

[the remark this answers is not in the document]
Comment (GrivIN) [reply db_processing/core/utils/file_utils.py]: pylint is wrong ;)
which one looks more clear? imo second version shows better when Connection parameters ends
```python
        with pysftp.Connection(
            hostname,
            username=username,
            password=password,
            port=port,
            cnopts=cnopts) as sftp:
            sftp.chdir(dirname)
            files = sftp.listdir_attr()
            files = [f for f in files if fnmatch.fnmatch(f.filename, filepath)]
            filename = sorted(files, key=lambda x: x.st_mtime, reverse=True)[0].filename
            sftp.getfo(filename, stream)
```
```python
        with pysftp.Connection(
                hostname,
                username=username,
                password=password,
                port=port,
                cnopts=cnopts) as sftp:
            sftp.chdir(dirname)
            files = sftp.listdir_attr()
            files = [f for f in files if fnmatch.fnmatch(f.filename, filepath)]
            filename = sorted(files, key=lambda x: x.st_mtime, reverse=True)[0].filename
            sftp.getfo(filename, stream)
```

File: db_processing/core/utils/file_utils.py [hunk 7]
```diff
@@ -1,76 +1,118 @@
 import logging
 import os
+import re
+import fnmatch
+import glob
+from urllib.parse import urlparse
 from io import BytesIO
 import gnupg
 import boto3
+import pysftp
```
Comment (GrivIN) [inline db_processing/core/utils/file_utils.py]: missing in requirements

**Pull request merged**